File: Sources/RavonCore/Dispatch/SwitchbackExperiment.swift
import Foundation

/// Experiment designs for evaluating a dispatch algorithm change.
///
/// ## Why this file exists
///
/// A/B testing a dispatch algorithm by randomising **individual orders** into treatment
/// and control is **invalid**, and the reason is structural rather than statistical:
/// orders in the same place at the same time draw from the *same courier pool*. Serving a
/// treated order well consumes a courier that a control order needed. The two arms
/// interfere, which violates the independence assumption every standard A/B test rests
/// on (SUTVA).
///
/// DoorDash's answer is the **switchback test**: randomise the algorithm over
/// **(geographic zone × time block)** cells, so an entire zone runs one variant for an
/// entire block and the market acts as its own comparison.
///
/// The value of doing this in a simulator is that **ground truth is knowable** — run the
/// world entirely on one algorithm, then entirely on the other, with the same seed, and
/// the difference is the true effect. That makes it possible to *measure the bias* of each
/// experiment design rather than reason about it in the abstract.
public enum ExperimentArm: String, Sendable, CaseIterable {
    case control
    case treatment
}

/// Assigns orders to arms. Deterministic, so an order's arm can be re-derived after the
/// simulation when computing per-arm metrics.
public struct ArmAssignment: Sendable {
    public let name: String
    private let assign: @Sendable (DispatchOrder) -> ExperimentArm
    private let assignRecord: @Sendable (MarketplaceSimulator.OrderRecord) -> ExperimentArm

    public init(
        name: String,
        assign: @escaping @Sendable (DispatchOrder) -> ExperimentArm,
        assignRecord: @escaping @Sendable (MarketplaceSimulator.OrderRecord) -> ExperimentArm
    ) {
        self.name = name
        self.assign = assign
        self.assignRecord = assignRecord
    }

    public func arm(for order: DispatchOrder) -> ExperimentArm { assign(order) }
    public func arm(for record: MarketplaceSimulator.OrderRecord) -> ExperimentArm {
        assignRecord(record)
    }

    /// Hash a UUID to a stable arm. Deterministic and independent of ordering, which is
    /// what makes the assignment re-derivable after the fact.
    private static func hashToArm(_ id: UUID, salt: UInt64) -> ExperimentArm {
        var hash: UInt64 = salt &+ 0x9E37_79B9_7F4A_7C15
        withUnsafeBytes(of: id.uuid) { bytes in
            for byte in bytes {
                hash = (hash ^ UInt64(byte)) &* 0x1000_0000_01B3
            }
        }
        return hash.multipliedReportingOverflow(by: 1).partialValue % 2 == 0 ? .control : .treatment
    }

    /// **The invalid design.** Each order independently coin-flipped into an arm, while
    /// both arms compete for one courier pool.
    public static func naiveOrderLevel(salt: UInt64 = 1) -> ArmAssignment {
        ArmAssignment(
            name: "naive-order-level-ab",
            assign: { hashToArm($0.id, salt: salt) },
            assignRecord: { hashToArm($0.id, salt: salt) }
        )
    }

    /// **The switchback design.** A whole (zone × time block) cell runs one variant, so
    /// orders that compete for the same couriers at the same time are almost always in
    /// the same arm.
    public static func switchback(
        grid: ZoneGrid, blockMinutes: Double, epoch: Date, salt: UInt64 = 1
    ) -> ArmAssignment {
        @Sendable func cellArm(zone: DispatchZone, block: Int) -> ExperimentArm {
            var hash = salt &+ 0x9E37_79B9_7F4A_7C15
            for value in [UInt64(bitPattern: Int64(zone.row)),
                          UInt64(bitPattern: Int64(zone.column)),
                          UInt64(bitPattern: Int64(block))] {
                hash = (hash ^ value) &* 0x1000_0000_01B3
                hash ^= hash >> 29
            }
            return hash % 2 == 0 ? .control : .treatment
        }
        return ArmAssignment(
            name: "switchback-zone-x-timeblock",
            assign: { order in
                let minutes = order.createdAt.timeIntervalSince(epoch) / 60
                let block = Int((minutes / blockMinutes).rounded(.down))
                return cellArm(zone: grid.zone(for: order.pickup), block: block)
            },
            assignRecord: { record in
                let block = Int((record.createdAtMinutes / blockMinutes).rounded(.down))
                return cellArm(zone: grid.zone(for: record.pickup), block: block)
            }
        )
    }
}

/// Runs two dispatchers side by side inside one simulated world, splitting orders by arm
/// while both draw from the **same courier pool** — which is precisely the interference
/// that makes naive A/B testing wrong.
public struct ExperimentDispatcher: Dispatcher {
    public let name: String
    public let control: any Dispatcher
    public let treatment: any Dispatcher
    public let assignment: ArmAssignment

    public init(control: any Dispatcher, treatment: any Dispatcher, assignment: ArmAssignment) {
        self.control = control
        self.treatment = treatment
        self.assignment = assignment
        self.name = "experiment[\(assignment.name)]"
    }

    public func assign(
        couriers: [DispatchCourier], orders: [DispatchOrder], now: Date
    ) -> [Assignment] {
        let treated = orders.filter { assignment.arm(for: $0) == .treatment }
        let controlled = orders.filter { assignment.arm(for: $0) == .control }

        // Whichever arm runs first gets first refusal on the courier pool. Alternating by
        // tick prevents that ordering from becoming a systematic advantage that would
        // masquerade as a treatment effect.
        let treatmentFirst = Int(now.timeIntervalSince1970 / 60) % 2 == 0

        var remaining = couriers
        var result: [Assignment] = []

        func runArm(_ dispatcher: any Dispatcher, _ subset: [DispatchOrder]) {
            guard !subset.isEmpty, !remaining.isEmpty else { return }
            let assignments = dispatcher.assign(couriers: remaining, orders: subset, now: now)
            let consumed = Set(assignments.map(\.courierID))
            remaining.removeAll { consumed.contains($0.id) }
            result.append(contentsOf: assignments)
        }

        if treatmentFirst {
            runArm(treatment, treated)
            runArm(control, controlled)
        } else {
            runArm(control, controlled)
            runArm(treatment, treated)
        }
        return result
    }
}

/// Measures a dispatch change under three designs and reports how badly each one lies.
public struct SwitchbackExperiment: Sendable {

    public struct ArmMetrics: Sendable {
        public let orders: Int
        public let assigned: Int
        public let meanDeliveryMinutes: Double
        public var assignmentRate: Double {
            orders == 0 ? 0 : Double(assigned) / Double(orders)
        }
    }

    public struct DesignResult: Sendable {
        public let designName: String
        public let control: ArmMetrics
        public let treatment: ArmMetrics
        /// Estimated effect on assignment rate, in percentage points.
        public var estimatedLiftPoints: Double {
            (treatment.assignmentRate - control.assignmentRate) * 100
        }
        /// Estimated effect on mean delivery time, in minutes (negative is better).
        public var estimatedDeliveryDeltaMinutes: Double {
            treatment.meanDeliveryMinutes - control.meanDeliveryMinutes
        }
    }

    public struct Report: Sendable {
        /// The true effect, from running the world entirely on each algorithm.
        public let groundTruthLiftPoints: Double
        public let groundTruthDeliveryDeltaMinutes: Double
        public let designs: [DesignResult]

        public func biasPoints(for designName: String) -> Double? {
            designs.first { $0.designName == designName }
                .map { $0.estimatedLiftPoints - groundTruthLiftPoints }
        }

        public var summary: String {
            var lines: [String] = []
            lines.append(String(
                format: "ground truth (full-world A vs full-world B):  lift %+.1f pts, delivery %+.1f min",
                groundTruthLiftPoints, groundTruthDeliveryDeltaMinutes
            ))
            for design in designs {
                let bias = design.estimatedLiftPoints - groundTruthLiftPoints
                // Pad in Swift: `%s` in String(format:) takes a C string, and handing it
                // a Swift String segfaults. `%@` is the correct specifier but does not
                // honour width flags.
                let label = design.designName.padding(
                    toLength: 30, withPad: " ", startingAt: 0
                )
                lines.append(
                    label + String(
                        format: "lift %+6.1f pts  (bias %+6.1f pts)  delivery %+6.1f min",
                        design.estimatedLiftPoints, bias, design.estimatedDeliveryDeltaMinutes
                    )
                )
            }
            return lines.joined(separator: "\n")
        }
    }

    private static func metrics(
        records: [MarketplaceSimulator.OrderRecord], arm: ExperimentArm, assignment: ArmAssignment
    ) -> ArmMetrics {
        let subset = records.filter { assignment.arm(for: $0) == arm }
        let delivered = subset.compactMap(\.totalDeliveryMinutes)
        return ArmMetrics(
            orders: subset.count,
            assigned: subset.filter { $0.assignedAtMinutes != nil }.count,
            meanDeliveryMinutes: delivered.isEmpty
                ? 0 : delivered.reduce(0, +) / Double(delivered.count)
        )
    }

    /// - Parameters:
    ///   - blockMinutes: switchback block length. Short blocks give more randomisation
    ///     units and more statistical power; long blocks reduce carryover between arms.
    ///     The trade DoorDash calls out explicitly.
    public static func run(
        config: MarketplaceSimulator.Config,
        control: any Dispatcher,
        treatment: any Dispatcher,
        zoneDivisions: Int = 3,
        blockMinutes: Double = 30
    ) -> Report {
        // Ground truth: two separate worlds, same seed, one algorithm each.
        let fullControl = MarketplaceSimulator.run(config: config, dispatcher: control)
        let fullTreatment = MarketplaceSimulator.run(config: config, dispatcher: treatment)
        let truthLift =
            (Double(fullTreatment.ordersAssigned) / Double(fullTreatment.ordersOffered)
             - Double(fullControl.ordersAssigned) / Double(fullControl.ordersOffered)) * 100
        let truthDelivery = fullTreatment.meanDeliveryMinutes - fullControl.meanDeliveryMinutes

        let epoch = Date(timeIntervalSince1970: 1_700_000_000)
        let grid = ZoneGrid(
            center: config.cityCenter, radiusKm: config.cityRadiusKm, divisions: zoneDivisions
        )

        let designs: [ArmAssignment] = [
            .naiveOrderLevel(salt: config.seed),
            .switchback(grid: grid, blockMinutes: blockMinutes, epoch: epoch, salt: config.seed),
        ]

        let results = designs.map { assignment -> DesignResult in
            let mixed = MarketplaceSimulator.run(
                config: config,
                dispatcher: ExperimentDispatcher(
                    control: control, treatment: treatment, assignment: assignment
                )
            )
            return DesignResult(
                designName: assignment.name,
                control: metrics(records: mixed.orderRecords, arm: .control, assignment: assignment),
                treatment: metrics(records: mixed.orderRecords, arm: .treatment, assignment: assignment)
            )
        }

        return Report(
            groundTruthLiftPoints: truthLift,
            groundTruthDeliveryDeltaMinutes: truthDelivery,
            designs: results
        )
    }
}

File: services/dispatch/src/main/kotlin/dev/ravon/dispatch/SwitchbackExperiment.kt
package dev.ravon.dispatch

import java.util.UUID

/**
 * Experiment designs for evaluating a dispatch algorithm change.
 *
 * Port of Sources/RavonCore/Dispatch/SwitchbackExperiment.swift.
 *
 * ## Why this exists
 *
 * A/B testing a dispatch algorithm by randomising **individual orders** is **invalid**,
 * and the reason is structural rather than statistical: orders in the same place at the
 * same time draw from the *same courier pool*. Serving a treated order well consumes a
 * courier a control order needed. The arms interfere, which violates the independence
 * assumption every standard A/B test rests on (SUTVA).
 *
 * The answer is the **switchback test**: randomise over **(zone × time block)** cells, so
 * an entire zone runs one variant for an entire block and the market acts as its own
 * comparison.
 *
 * Doing this in a simulator is what makes it checkable — run the world entirely on one
 * algorithm, then entirely on the other with the same seed, and the difference is the
 * *true* effect. So the bias of each design can be **measured** rather than argued about.
 */
enum class ExperimentArm { CONTROL, TREATMENT }

/**
 * Assigns orders to arms. Deterministic, so an order's arm can be re-derived after the
 * simulation when computing per-arm metrics.
 */
class ArmAssignment(
    val name: String,
    private val armForOrder: (DispatchOrder) -> ExperimentArm,
    private val armForRecord: (MarketplaceSimulator.OrderRecord) -> ExperimentArm,
) {
    fun arm(order: DispatchOrder): ExperimentArm = armForOrder(order)
    fun arm(record: MarketplaceSimulator.OrderRecord): ExperimentArm = armForRecord(record)

    companion object {
        /**
         * Hash a UUID to a stable arm — FNV-style over the 16 identity bytes, matching
         * the Swift original's `withUnsafeBytes(of: id.uuid)` byte order (big-endian:
         * most-significant bits first).
         */
        private fun hashToArm(id: UUID, salt: ULong): ExperimentArm {
            var hash: ULong = salt + 0x9E3779B97F4A7C15UL
            val bytes = ByteArray(16)
            var msb = id.mostSignificantBits
            var lsb = id.leastSignificantBits
            for (i in 7 downTo 0) { bytes[i] = (msb and 0xFF).toByte(); msb = msb ushr 8 }
            for (i in 15 downTo 8) { bytes[i] = (lsb and 0xFF).toByte(); lsb = lsb ushr 8 }
            for (b in bytes) {
                hash = (hash xor (b.toULong() and 0xFFUL)) * 0x100000001B3UL
            }
            return if (hash % 2UL == 0UL) ExperimentArm.CONTROL else ExperimentArm.TREATMENT
        }

        /**
         * **The invalid design.** Each order independently coin-flipped, while both arms
         * compete for one courier pool.
         *
         * A caveat the Swift original cannot state about itself: there, order ids come
         * from `UUID()` and are therefore *random per run*, so this design's arm split —
         * and every number derived from it — is not reproducible. This port derives ids
         * from the order index, so the split is deterministic. That makes the Kotlin
         * naive-design figures stable but **not** comparable to any particular Swift run.
         * Only the *property* (that this design is badly biased) transfers, which is
         * exactly what the tests assert.
         */
        fun naiveOrderLevel(salt: ULong = 1UL): ArmAssignment = ArmAssignment(
            name = "naive-order-level-ab",
            armForOrder = { hashToArm(it.id, salt) },
            armForRecord = { hashToArm(it.id, salt) },
        )

        /**
         * **The switchback design.** A whole (zone × time block) cell runs one variant,
         * so orders competing for the same couriers at the same time are almost always in
         * the same arm. Deterministic in both languages.
         */
        fun switchback(
            grid: ZoneGrid,
            blockMinutes: Double,
            salt: ULong = 1UL,
        ): ArmAssignment {
            fun cellArm(zone: DispatchZone, block: Int): ExperimentArm {
                var hash = salt + 0x9E3779B97F4A7C15UL
                for (value in longArrayOf(zone.row.toLong(), zone.column.toLong(), block.toLong())) {
                    hash = (hash xor value.toULong()) * 0x100000001B3UL
                    hash = hash xor (hash shr 29)
                }
                return if (hash % 2UL == 0UL) ExperimentArm.CONTROL else ExperimentArm.TREATMENT
            }
            return ArmAssignment(
                name = "switchback-zone-x-timeblock",
                armForOrder = { order ->
                    val minutes =
                        (order.createdAtEpochSeconds - MarketplaceSimulator.EPOCH_SECONDS) / 60
                    cellArm(grid.zone(order.pickup), StrictMath.floor(minutes / blockMinutes).toInt())
                },
                armForRecord = { record ->
                    cellArm(
                        grid.zone(record.pickup),
                        StrictMath.floor(record.createdAtMinutes / blockMinutes).toInt(),
                    )
                },
            )
        }
    }
}

/**
 * Runs two dispatchers side by side inside one simulated world, splitting orders by arm
 * while both draw from the **same courier pool** — which is precisely the interference
 * that makes naive A/B testing wrong.
 */
class ExperimentDispatcher(
    private val control: Dispatcher,
    private val treatment: Dispatcher,
    private val assignment: ArmAssignment,
) : Dispatcher {
    override val name: String = "experiment[${assignment.name}]"

    override fun assign(
        couriers: List<DispatchCourier>,
        orders: List<DispatchOrder>,
        nowEpochSeconds: Double,
    ): List<Assignment> {
        val treated = orders.filter { assignment.arm(it) == ExperimentArm.TREATMENT }
        val controlled = orders.filter { assignment.arm(it) == ExperimentArm.CONTROL }

        // Whichever arm runs first gets first refusal on the courier pool. Alternating by
        // tick stops that ordering from becoming a systematic advantage that would
        // masquerade as a treatment effect.
        //
        // The original reads `Int(now.timeIntervalSince1970 / 60) % 2`, i.e. the *Unix*
        // view of the instant, so the parity depends on the absolute epoch and not just
        // the elapsed time. Hence the conversion rather than using minutes directly.
        val unixSeconds =
            nowEpochSeconds - MarketplaceSimulator.EPOCH_SECONDS + MarketplaceSimulator.EPOCH_UNIX_SECONDS
        val treatmentFirst = (unixSeconds / 60).toLong() % 2L == 0L

        val remaining = couriers.toMutableList()
        val result = mutableListOf<Assignment>()

        fun runArm(dispatcher: Dispatcher, subset: List<DispatchOrder>) {
            if (subset.isEmpty() || remaining.isEmpty()) return
            val assignments = dispatcher.assign(remaining, subset, nowEpochSeconds)
            val consumed = assignments.map { it.courierId }.toSet()
            remaining.removeAll { it.id in consumed }
            result += assignments
        }

        if (treatmentFirst) {
            runArm(treatment, treated)
            runArm(control, controlled)
        } else {
            runArm(control, controlled)
            runArm(treatment, treated)
        }
        return result
    }
}

/** Measures a dispatch change under two designs and reports how badly each one lies. */
object SwitchbackExperiment {

    data class ArmMetrics(
        val orders: Int,
        val assigned: Int,
        val meanDeliveryMinutes: Double,
    ) {
        val assignmentRate: Double get() = if (orders == 0) 0.0 else assigned.toDouble() / orders
    }

    data class DesignResult(
        val designName: String,
        val control: ArmMetrics,
        val treatment: ArmMetrics,
    ) {
        /** Estimated effect on assignment rate, in percentage points. */
        val estimatedLiftPoints: Double
            get() = (treatment.assignmentRate - control.assignmentRate) * 100

        /** Estimated effect on mean delivery time, in minutes (negative is better). */
        val estimatedDeliveryDeltaMinutes: Double
            get() = treatment.meanDeliveryMinutes - control.meanDeliveryMinutes
    }

    data class Report(
        /** The true effect, from running the world entirely on each algorithm. */
        val groundTruthLiftPoints: Double,
        val groundTruthDeliveryDeltaMinutes: Double,
        val designs: List<DesignResult>,
    ) {
        fun biasPoints(designName: String): Double? =
            designs.firstOrNull { it.designName == designName }
                ?.let { it.estimatedLiftPoints - groundTruthLiftPoints }

        val summary: String
            get() = buildList {
                add(
                    "ground truth (full-world A vs full-world B):  lift %+.1f pts, delivery %+.1f min"
                        .format(groundTruthLiftPoints, groundTruthDeliveryDeltaMinutes)
                )
                designs.forEach { design ->
                    val bias = design.estimatedLiftPoints - groundTruthLiftPoints
                    // `%-30s` is safe here. The Swift original had to pad by hand because
                    // `String(format:)` with `%s` takes a C string and segfaults on a
                    // Swift String; `%@` is correct but ignores width flags. The JVM has
                    // no such trap.
                    add(
                        "%-30s lift %+6.1f pts  (bias %+6.1f pts)  delivery %+6.1f min".format(
                            design.designName, design.estimatedLiftPoints, bias,
                            design.estimatedDeliveryDeltaMinutes,
                        )
                    )
                }
            }.joinToString("\n")
    }

    private fun metrics(
        records: List<MarketplaceSimulator.OrderRecord>,
        arm: ExperimentArm,
        assignment: ArmAssignment,
    ): ArmMetrics {
        val subset = records.filter { assignment.arm(it) == arm }
        val delivered = subset.mapNotNull { it.totalDeliveryMinutes }
        return ArmMetrics(
            orders = subset.size,
            assigned = subset.count { it.assignedAtMinutes != null },
            meanDeliveryMinutes =
                if (delivered.isEmpty()) 0.0
                else delivered.fold(0.0) { a, b -> a + b } / delivered.size,
        )
    }

    /**
     * @param blockMinutes switchback block length. Short blocks give more randomisation
     *   units and more statistical power; long blocks reduce carryover between arms. The
     *   trade DoorDash calls out explicitly.
     */
    fun run(
        config: MarketplaceSimulator.Config,
        control: Dispatcher,
        treatment: Dispatcher,
        zoneDivisions: Int = 3,
        blockMinutes: Double = 30.0,
    ): Report {
        // Ground truth: two separate worlds, same seed, one algorithm each.
        val fullControl = MarketplaceSimulator.run(config, control)
        val fullTreatment = MarketplaceSimulator.run(config, treatment)
        val truthLift = (
            fullTreatment.ordersAssigned.toDouble() / fullTreatment.ordersOffered -
                fullControl.ordersAssigned.toDouble() / fullControl.ordersOffered
            ) * 100
        val truthDelivery = fullTreatment.meanDeliveryMinutes - fullControl.meanDeliveryMinutes

        val grid = ZoneGrid(config.cityCenter, config.cityRadiusKm, zoneDivisions)
        val designs = listOf(
            ArmAssignment.naiveOrderLevel(salt = config.seed),
            ArmAssignment.switchback(grid, blockMinutes, salt = config.seed),
        )

        val results = designs.map { assignment ->
            val mixed = MarketplaceSimulator.run(
                config,
                ExperimentDispatcher(control, treatment, assignment),
            )
            DesignResult(
                designName = assignment.name,
                control = metrics(mixed.orderRecords, ExperimentArm.CONTROL, assignment),
                treatment = metrics(mixed.orderRecords, ExperimentArm.TREATMENT, assignment),
            )
        }

        return Report(truthLift, truthDelivery, results)
    }
}
